inspector: introduce a smoke test for inspector

eugeneo opened the pull request:
<!--
Thank you for your pull request. Please review below requirements.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test nosign` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] commit message follows commit guidelines

##### Affected core subsystem(s)
Adds a new test case that checks the inspector


##### Description of change
<!-- Provide a description of the change below this comment. -->

This test executes a simple debug session over the inspector protocol.

CC: @ofrobots

Trott commented:
Assuming the test is expected to work on Windows, I imagine there should probably be a  `test-inspector` task added to `vcbuild.bat` too.

Trott commented:
`test-inspector` takes only a second or so to run and doesn't seem to reach out to the internet or anything like that. Any reason not to add the `inspector` dir to the `test` and `test-ci` tasks?

Trott commented:
@nodejs/testing

mhdawson commented:
It would be good to have the  added, provided it passes across the platforms

Trott commented:
Nit: add an entry for the `inspector` directory to the `test/README.md` file.

(We didn't do that for `pseudo-tty`. If @Fishrock123 doesn't want to do that that would be a good first contribution for anyone following along. ❤️ )

ofrobots commented on src/inspector_socket.cc:
```diff
@@ -33,7 +33,7 @@ static void dump_hex(const char* buf, size_t len) {
     cptr = ptr;
     for (i = 0; i < 16 && ptr < end; i++) {
-      printf("%2.2X  ", *(ptr++));
+      printf("%2.2X  ", (unsigned char) *(ptr++));
```
Avoid C style casts.

eugeneo commented:
Adressed the comments:

- Added tests to test and test-ci
- Updated vcbuild.bat
- Added an entry to README.md
- Replaced the cast.

Trott commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/3970/

santigimeno commented on test/inspector/test-inspector.js:
```diff
@@ -0,0 +1,167 @@
+'use strict';
+const assert = require('assert');
+require('../common');
```
I think `../common` should be the first require

santigimeno commented on test/inspector/test-inspector.js:
```diff
@@ -0,0 +1,167 @@
+
+function setupExpectBreakOnLine(line, url, session, scopeIdCallback) {
+  return function(message) {
+    if ('Debugger.paused' == message['method']) {
```
maybe s/==/===

Trott commented:
Looks like this is coming up from time to time on Linux variants:

```
not ok 1166 inspector/test-inspector
# AssertionError: Timed out waiting for Waiting for the debugger to disconnect...
#     at Timeout.setTimeout (/home/iojs/build/workspace/node-test-commit-linux/nodes/centos6-64/test/inspector/inspector-helper.js:235:9)
#     at tryOnTimeout (timers.js:232:11)
#     at Timer.listOnTimeout (timers.js:202:5)
# Verifying debugger stops on start (--debug-brk option)
# [err] Debugger attached.
# 
# Setting a breakpoint and verifying it is hit
# [out] A message 5
# 
# Verify we can read current application state
# Verify node waits for the frontend to disconnect
# [out] Outputed message #1
# 
# [err] Waiting for the debugger to disconnect...
# 
  ---
  duration_ms: 15.543
```

santigimeno commented on test/fixtures/loop.js:
```diff
@@ -0,0 +1,10 @@
+var k = 1;
+console.log('A message', 5);
+while (t > 0) {
+  if (t++ == 1000) {
```
s/==/===

eugeneo replied:
Thanks for pointing out - I haven't noticed that comment. Fixed.

santigimeno commented on test/inspector/inspector-helper.js:
```diff
@@ -0,0 +1,429 @@
+    dataLen = buffer.readUInt16BE(2);
+    bodyOffset = 4;
+  } else if (dataLen == 127) {
+    dataLen = buffer.readUInt32BE(2);
```
s/==/=== also here and in some other places

santigimeno commented on test/inspector/inspector-helper.js:
```diff
@@ -0,0 +1,429 @@
+
+  for (let i = 0; i < message_buf.length; i++) {
+    message_buf[i] = message_buf[i] ^ (1 << (i % 4));
+  }
```
Not too sure if it's enforced in all the codebase but I think one-line for's and if's don't use brackets.

cjihrig commented on test/fixtures/loop.js:
```diff
@@ -0,0 +1,10 @@
+    console.log('Outputed message #' + k++);
+  }
+}
+process.exit (55);
```
nit: no space between `exit` and `(`

cjihrig commented on test/inspector/inspector-helper.js:
```diff
@@ -0,0 +1,412 @@
+'use strict';
+const assert = require('assert');
+const common = require('../common');
```
Please make this the first thing required.

cjihrig commented on test/inspector/inspector-helper.js:
```diff
@@ -0,0 +1,412 @@
+const TIMEOUT = 15 * 1000;
+
+const mainScript = path.join(
+  common.fixturesDir,
```
I think this `path.join()` will all fit on a single line.

cjihrig commented on test/inspector/inspector-helper.js:
```diff
@@ -0,0 +1,412 @@
+function send(socket, message, id, callback) {
+  const msg = JSON.parse(JSON.stringify(message)); // Clone!
+  msg['id'] = id;
+  const message_buf = Buffer.from(JSON.stringify(msg));
```
Pleas use camelCase in JS here and throughout.

cjihrig commented on test/inspector/inspector-helper.js:
```diff
@@ -0,0 +1,412 @@
+
+  for (let i = 0; i < message_buf.length; i++)
+    message_buf[i] = message_buf[i] ^ (1 << (i % 4));
+  socket.write(
```
Does the linter complain about indentation here?

eugeneo replied:
I'm doind make -j4 test before each upload, linter does not complain.

cjihrig commented on test/inspector/inspector-helper.js:
```diff
@@ -0,0 +1,412 @@
+function processIncomingMessage(socket, buffer, handler) {
+  if (buffer.length < 2)
+    return 0;
+  assert.equal(0x81, buffer[0]);
```
Can you change the `assert.equal()` calls to `assert.strictEqual()`.

cjihrig commented on test/inspector/inspector-helper.js:
```diff
@@ -0,0 +1,412 @@
+}
+
+function tearDown(child, err) {
+  child && child.kill();
```
Is it necessary to check if the child exists? Shouldn't we be able to guarantee that in our tests?

cjihrig commented on test/inspector/inspector-helper.js:
```diff
@@ -0,0 +1,412 @@
+}
+
+function checkHttpResponse(port, path, callback) {
+  http.get({port: port, path: path}, function(r) {
```
Feel free to ignore this, but you could shorten this to `{port, path}`.

eugeneo commented:
Thanks for all the comments. Updated the CL:

- Addressed a race condition when "Waiting for the debugger to disconnect" arrives before the test set the watcher.
- s/==/===/ and s/!=/!==/
- Removed brackets around single-line statements (retained the brackets if another branch of the same if statement has it.
- Did other changes suggested by @cjihrig

santigimeno commented on test/inspector/inspector-helper.js:
```diff
@@ -0,0 +1,406 @@
+      [ '--inspect', '--debug-brk', mainScript ]);
+
+  const timeoutId = setTimeout(() => {
+    assert(false, 'Child process did not start properly');
```
may use `common.fail` here

eugeneo replied:
Done

santigimeno commented on test/inspector/inspector-helper.js:
```diff
@@ -0,0 +1,406 @@
+  child.stderr.once('data', (data) => {
+    clearTimeout(timeoutId);
+    const text = data.toString('utf8');
+    const match = text.match(/Debugger listening on port (\d+)/);
```
can't it be possible that the `Debugger listening ...` text isn't present after only one `data` event? I think it could be split in multiple events.

eugeneo replied:
Realistically, I doubt that :) But I introduced accumulator code just in case.

santigimeno commented on test/inspector/inspector-helper.js:
```diff
@@ -0,0 +1,406 @@
+        });
+      }
+      enqueue(tests);
+    }).on('response', () => assert(false, 'Upgrade was not received'));
```
use `common.fail`?

eugeneo replied:
Done.

santigimeno commented on test/inspector/inspector-helper.js:
```diff
@@ -0,0 +1,406 @@
+  this.enqueue_((callback) => {
+    if (this.running_) {
+      const timeoutId =
+          setTimeout(() => assert(false, 'Have not terminated'),
```
and here

eugeneo replied:
Done.

santigimeno commented on test/inspector/inspector-helper.js:
```diff
@@ -0,0 +1,406 @@
+  return this.enqueue((cb) => {
+    if (done) {
+      cb();
+    } else {
```
Style nit: maybe `return cb();` here to get rid of one level of indentation.

eugeneo replied:
Redid the code quite a bit here.

santigimeno commented on test/inspector/inspector-helper.js:
```diff
@@ -0,0 +1,406 @@
+      cb();
+    } else {
+      const timeoutId = setTimeout(
+          () => assert(false, 'Response was not received: ' +
```
`setTimeout(()=>` as in the rest of the file?

eugeneo replied:
Introduced a function to avoid repetitions.

santigimeno commented on test/inspector/inspector-helper.js:
```diff
@@ -0,0 +1,406 @@
+  process.on('exit', (code, signal) => {
+    assert(this.childInstanceDone, 'Child instance died prematurely');
+    this.returnCode_ = code;
+    this.running_ = false;
```
Maybe it's a good idea to use `close` instead of `exit` to be sure all the data from the stderr stream has been received.

eugeneo replied:
Done.

eugeneo commented:
Thank you for the review. I uploaded a new version, please take another look.

ofrobots commented:
New CI: https://ci.nodejs.org/job/node-test-pull-request/4028/

ofrobots commented:
The test times out on [windows](https://ci.nodejs.org/job/node-test-binary-windows/3760/RUN_SUBSET=1,VS_VERSION=vcbt2015,label=win10/tapTestReport/test.tap-293/):

```
# TIMEOUT
# AssertionError: Child process did not start properly
# at Object.exports.fail (c:\workspace\node-test-binary-windows\RUN_SUBSET\1\VS_VERSION\vcbt2015\label\win10\test\common.js:438:10)
# at Timeout.setTimeout (c:\workspace\node-test-binary-windows\RUN_SUBSET\1\VS_VERSION\vcbt2015\label\win10\test\inspector\inspector-helper.js:102:34)
# at tryOnTimeout (timers.js:232:11)
# at Timer.listOnTimeout (timers.js:202:5)
```

eugeneo commented:
Windows seems to buffer all stderr output, I had to add fflush to inspector code

bnoordhuis commented on src/inspector_agent.cc:
```diff
@@ -620,6 +622,7 @@ bool AgentImpl::OnInspectorHandshakeIO(inspector_socket_t* socket,
     return false;
   default:
     UNREACHABLE();
+    return false;
```
Unrelated change.  Is it to squelch a compiler warning?

eugeneo replied:
Seems to be a VC++ warning. I fixed it as I was testing on Windows.

ofrobots commented:
New CI: https://ci.nodejs.org/job/node-test-pull-request/4049/

eugeneo commented on test/inspector/inspector-helper.js:
```diff
@@ -0,0 +1,425 @@
+      callback);
+}
+
+function processIncomingMessage(socket, buffer, handler) {
```
Thanks. Fixed.

ofrobots replied:
Did you fix this? Either the new github review mechanism is lying to me, or socket is still a parameter here.

eugeneo replied:
Not sure what I did wrong - but I reuploaded the fix.

eugeneo commented:
Addressed the nit. Please take another look.

ofrobots commented:
Relaunched CI: https://ci.nodejs.org/job/node-test-pull-request/4129/

ofrobots commented:
Thanks. Landed as b4a249f.

Trott commented:
Just got a failure for this test on CI: https://ci.nodejs.org/job/node-test-commit-linux/5205/nodes=centos5-32/tapTestReport/test.tap-1174/

```
not ok 1174 inspector/test-inspector
# AssertionError: Have not terminated
# at Object.exports.fail (/home/iojs/build/workspace/node-test-commit-linux/nodes/centos5-32/test/common.js:442:10)
# at Timeout.setTimeout (/home/iojs/build/workspace/node-test-commit-linux/nodes/centos5-32/test/inspector/inspector-helper.js:102:34)
# at tryOnTimeout (timers.js:232:11)
# at Timer.listOnTimeout (timers.js:202:5)
# [err] Debugger listening on port 9229.
# [err] Warning: This is an experimental feature and could change at any time.
# [err] To start debugging, open the following URL in Chrome:
# [err]     chrome-devtools://devtools/remote/serve_file/@60cd6e859b9f557d2312f5bf532f6aec5f284980/inspector.html?experiments=true&v8only=true&ws=localhost:9229/node
# [err] 
# [test] Verifying debugger stops on start (--debug-brk option)
# [err] Debugger attached.
# [err] 
# [test] Setting a breakpoint and verifying it is hit
# [out] A message 5
# [out] 
# [test] Verify we can read current application state
# [test] Verify node waits for the frontend to disconnect
# [out] Outputed message #1
# [out] 
# [err] Waiting for the debugger to disconnect...
# [err] 
```

Trott commented:
Stress test: https://ci.nodejs.org/job/node-stress-single-test/921/nodes=centos5-32/console

eugeneo commented:
@Trott How do I run this test locally? Looking at the output, it could be that this is simply a timeout while the node is shutting down. Are there any standards for how long the test can wait on failure?

Trott commented:
@eugeneo `python tools/test.py -p tap --mode=release inspector/test-inspector`

On the stress test, it looks like it's failing about one out of every 50 runs on CentOS 5 (32-bit). So you'll probably want to put that in a loop or something.

eugeneo commented:
It is not failing on my Ubuntu 64 workstation. Can you try this commit -
https://github.com/eugeneo/node/commit/542c02348a5f2236f0c98c719067790a0a0d6332
on
that environment? I added some more debug output.

On Mon, Sep 19, 2016 at 4:53 PM Rich Trott <notifications@github.com> wrote:

> @eugeneo <https://github.com/eugeneo> python tools/test.py -p tap
> --mode=release inspector/test-inspector
>
> On the stress test, it looks like it's failing about one out of every 50
> runs on CentOS 5 (32-bit). So you'll probably want to put that in a loop or
> something.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/nodejs/node/pull/8429#issuecomment-248162862>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AARkre3qQ_CjIRfaFI-Oizo6DfTb7mzfks5qryB0gaJpZM4J2cMg>
> .
>

Trott commented:
Here you go. This time, the test timed out after 8 successful runs.
¯\\_(ツ)_/¯
https://ci.nodejs.org/job/node-stress-single-test/922/nodes=centos5-32/console

```
not ok 1 inspector/test-inspector
# TIMEOUT
# [err] Debugger listening on port 9229.
# [err] Warning: This is an experimental feature and could change at any time.
# [err] To start debugging, open the following URL in Chrome:
# [err]     chrome-devtools://devtools/remote/serve_file/@60cd6e859b9f557d2312f5bf532f6aec5f284980/inspector.html?experiments=true&v8only=true&ws=localhost:9229/node
# [err] 
# [test] Verifying debugger stops on start (--debug-brk option)
# [err] Debugger attached.
# [err] 
# [test] Setting a breakpoint and verifying it is hit
# [out] A message 5
# [out] 
# [test] Verify we can read current application state
# [test] Verify node waits for the frontend to disconnect
# [out] Outputed message #1
# [out] 
# [err] Waiting for the debugger to disconnect...
# [err] 
# [test] Connection terminated
```

Trott commented:
I should mention that the timeout is enforced by `test.py` and is 60 seconds.

eugeneo commented:
Thank you for looking into this.

I still have no idea where this failure is coming from... What I see in the trace is that the test does not get notification when the child process exits. I do not know if the child process exits. I uploaded a new commit - https://github.com/eugeneo/node/commit/972120d1dcd1d6c0913c56a98f7048b7d3cb8466 - that should indicate if the child exits. Can we try it on CentOS.

Note that this commit adds some tracing and thus breaks some tests other then test-inspector.

Is there a way for me to access one of the machines where the failure occurs?

Trott commented:
> Is there a way for me to access one of the machines where the failure occurs?

Not sure what the process is, but that would be something handled by the Build working group. @nodejs/build

Trott commented:
I queued up the test in Jenkins but it's waiting on a host to become available, I think....  Coming soon, though...

Trott commented:
Test is running... https://ci.nodejs.org/job/node-stress-single-test/925/nodes=centos5-32/console

eugeneo commented:
Thank you! This was most helpful - looks like the child node does not actually exit, which is a bug. I'm looking into it.

Trott commented:
Because the Jenkins results get cleared out from time to time, here's the result, in case someone needs to refer back to it later:

```
# TIMEOUT
# [err] Debugger listening on port 9229.
# [err] Warning: This is an experimental feature and could change at any time.
# [err] To start debugging, open the following URL in Chrome:
# [err]     chrome-devtools://devtools/remote/serve_file/@60cd6e859b9f557d2312f5bf532f6aec5f284980/inspector.html?experiments=true&v8only=true&ws=localhost:9229/node
# [err] 
# [test] Verifying debugger stops on start (--debug-brk option)
# [err] Debugger attached.
# [err] 
# [test] Setting a breakpoint and verifying it is hit
# [out] A message 5
# [out] 
# [test] Verify we can read current application state
# [test] Verify node waits for the frontend to disconnect
# [out] Outputed message #1
# [out] 
# [err] Waiting for the debugger to disconnect...
# [err] 
# [test] Connection terminated
# [err] [../src/inspector_agent.cc:781] Frontend's gone, can exit
# [err] 
# [err] [../src/inspector_agent.cc:790] Frontend's disconnect handled
# [err] 
```

eugeneo commented:
https://github.com/nodejs/node/issues/8669 was created for this issue, I will be posting patches there.

gibfahn commented:
@Trott @eugeneo FYI, to repeat tests with the test runner you can just do:
```
python tools/test.py -p tap --repeat 100 --mode=release inspector/test-inspector
```

eugeneo commented:
@gibfahn I cannot reproduce it outside of the CI - I just ran it again on my Ubuntu 64 system...
